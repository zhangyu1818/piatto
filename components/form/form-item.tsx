import * as React from 'react'
import { useController } from 'react-hook-form'
import classNames from 'classnames'

import type { RuleItem } from 'hook-form-async-validator'

import useConfig from '../hooks/useConfig'

export interface FormItemProps extends React.HTMLAttributes<HTMLDivElement> {
  name: string
  children: React.ReactElement
  rules?: RuleItem | RuleItem[]
}

const FormItem: React.FC<FormItemProps> = ({ name, children, className, rules, ...restProps }) => {
  const { getPrefixCls } = useConfig()
  const prefixCls = getPrefixCls('form-item')

  const { field, fieldState } = useController({
    name,
    defaultValue: '',
  })

  const { error } = fieldState

  const classes = classNames(prefixCls, className, {
    [`${prefixCls}-error`]: error,
  })

  return (
    <div className={classes}  {...restProps}>
      {React.cloneElement(children, { ...field, name })}
      <p className={`${prefixCls}-explain`}>{error ?? ''}</p>
    </div>
  )
}

export default FormItem
