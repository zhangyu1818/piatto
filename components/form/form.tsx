import * as React from 'react'
import { useForm, FormProvider } from 'react-hook-form'
import resolver from 'hook-form-async-validator'

import devWarning from '../utils/dev-warning'

import type {
  UseFormReturn,
  SubmitHandler,
  SubmitErrorHandler,
  FieldValues,
  DeepPartial,
  UnpackNestedValue,
} from 'react-hook-form/dist/types'
import type { Rules } from 'hook-form-async-validator'

export interface FormProps extends Omit<React.HTMLAttributes<HTMLFormElement>, 'onSubmit'> {
  form?: UseFormReturn
  onFinish?: SubmitHandler<FieldValues>
  onError?: SubmitErrorHandler<FieldValues>
  rules?: Rules
  defaultValues?: Partial<UnpackNestedValue<DeepPartial<FieldValues>>>
  children: React.ReactElement | React.ReactElement[]
}

const noop = () => {}

const InternalForm: React.ForwardRefRenderFunction<unknown, FormProps> = (
  {
    children,
    form: propsForm,
    onFinish = noop,
    onError,
    rules: propsRules,
    defaultValues,
    ...restProps
  },
  ref,
) => {
  devWarning(
    propsForm && (propsRules || defaultValues),
    `${propsRules ? 'rules' : ''} ${
      defaultValues ? 'defaultValues' : ''
    } is not valid when form is provided,please pass rules to your form instance`,
  )

  const formRules = React.useMemo(() => {
    const mergeRules = { ...propsRules } as Rules
    React.Children.forEach<React.ReactElement>(children, (item) => {
      const { name, rules } = item.props
      if (rules && name) mergeRules[name] = rules
    })
    return mergeRules
  }, [propsRules, children])

  const form = useForm({
    mode: 'onTouched',
    resolver: resolver(formRules),
    defaultValues,
  })

  const formInstance = propsForm || form

  React.useImperativeHandle(ref, () => formInstance)

  const { handleSubmit } = formInstance

  return (
    <FormProvider {...formInstance}>
      <form {...restProps} onSubmit={handleSubmit(onFinish, onError)}>
        {children}
      </form>
    </FormProvider>
  )
}

const Form = React.forwardRef<UseFormReturn, FormProps>(InternalForm) as (
  props: React.PropsWithChildren<FormProps> & { ref?: React.Ref<UseFormReturn> },
) => React.ReactElement

export type { UseFormReturn }
export default Form
