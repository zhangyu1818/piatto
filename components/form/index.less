@import '../styles/theme';
@import '../styles/utils/textOverflow';

@form-prefix-class: ~'@{prefix-class}-form';
@input-prefix-class: ~'@{prefix-class}-input';

.@{form-prefix-class}-item {
  // error
  &.@{form-prefix-class}-item-error {
    // input error
    .@{input-prefix-class} {
      border-color: @border-color-error;
      &:focus-within {
        box-shadow: 0 0 0 @box-shadow-spread fade(@border-color-error, 20%);
      }
    }
  }

  // error label
  &-explain {
    display: flex;
    align-items: center;
    height: @form-item-explain-height;
    margin: 0 @margin-sm;
    color: @form-item-explain-error-color;
    font-size: @form-item-explain-font-size;
    line-height: @form-item-explain-height;
    .textOverflow();
  }
}
