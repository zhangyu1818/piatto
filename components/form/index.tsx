export { default as Form } from 'rc-field-form';
export { default as FormItem } from './form-item';
