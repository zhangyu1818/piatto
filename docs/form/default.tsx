import React from 'react';
import { Form, Input, Button } from 'piatto';
import { UserOutlined, PhoneOutlined, CheckCircleOutlined } from '@ant-design/icons';

const DependenceInput: React.FC = props => {
  const phone = Form.useWatch({ name: 'phone', defaultValue: '' });
  const isValid = /^\d{11}$/.test(phone);
  return (
    <Input.GetCode
      {...props}
      placeholder="code"
      addonBefore={<CheckCircleOutlined />}
      block
      buttonText="Get Code"
      buttonDisabled={!isValid}
      onGetCode={() => new Promise(resolve => setTimeout(resolve, 1500))}
    />
  );
};

export default () => (
  <Form
    onFinish={values => {
      // eslint-disable-next-line no-console
      console.log(values);
    }}
  >
    <Form.Item name="username" rules={[{ required: true, message: 'place input your username' }]}>
      <Input placeholder="username" addonBefore={<UserOutlined />} block />
    </Form.Item>
    <Form.Item
      name="phone"
      rules={[
        { required: true, message: 'place input your phone' },
        { pattern: /^1\d{10}$/, message: 'invalid number' },
      ]}
    >
      <Input placeholder="phone(11)" addonBefore={<PhoneOutlined />} block />
    </Form.Item>
    <Form.Item name="code" rules={[{ required: true, message: 'place input your code' }]}>
      <DependenceInput />
    </Form.Item>
    <Button htmlType="submit">Submit</Button>
    <p style={{ fontSize: '0.33rem', margin: 0 }}>
      if phone number is invalid, get code button will disabled
    </p>
  </Form>
);
