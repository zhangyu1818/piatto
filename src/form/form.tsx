import React, { HTMLAttributes, useImperativeHandle, useMemo } from 'react';
import { useForm, FormProvider } from 'react-hook-form';
import resolver, { Rules } from 'hook-form-async-validator';
import type {
  UseFormMethods,
  SubmitHandler,
  SubmitErrorHandler,
  FieldValues,
  DeepPartial,
  UnpackNestedValue,
} from 'react-hook-form/dist/types';

import devWarning from '../utils/dev-warning';

export interface FormProps extends Omit<HTMLAttributes<HTMLFormElement>, 'onSubmit'> {
  form?: UseFormMethods;
  onFinish?: SubmitHandler<FieldValues>;
  onError?: SubmitErrorHandler<FieldValues>;
  rules?: Rules;
  defaultValues?: Partial<UnpackNestedValue<DeepPartial<FieldValues>>>;
  children: React.ReactElement | React.ReactElement[];
}

const noop = () => {};

const InternalForm: React.ForwardRefRenderFunction<unknown, FormProps> = (
  {
    children,
    form: propsForm,
    onFinish = noop,
    onError,
    rules: propsRules,
    defaultValues,
    ...restProps
  },
  ref,
) => {
  devWarning(
    propsForm && (propsRules || defaultValues),
    `${propsRules ? 'rules' : ''} ${
      defaultValues ? 'defaultValues' : ''
    } is not valid when form is provided,please pass rules to your form instance`,
  );

  const formRules = useMemo(() => {
    const mergeRules = { ...propsRules } as Rules;
    React.Children.forEach<React.ReactElement>(children, (item) => {
      const { name, rules } = item.props;
      if (rules && name) mergeRules[name] = rules;
    });
    return mergeRules;
  }, [propsRules, children]);

  const form = useForm({
    mode: 'onTouched',
    resolver: resolver(formRules),
    defaultValues,
  });

  const formInstance = propsForm || form;

  useImperativeHandle(ref, () => formInstance);

  const { handleSubmit } = formInstance;

  return (
    <FormProvider {...formInstance}>
      <form {...restProps} onSubmit={handleSubmit(onFinish, onError)}>
        {children}
      </form>
    </FormProvider>
  );
};

const Form = React.forwardRef<UseFormMethods, FormProps>(InternalForm) as (
  props: React.PropsWithChildren<FormProps> & { ref?: React.Ref<UseFormMethods> },
) => React.ReactElement;

export type { UseFormMethods };
export default Form;
