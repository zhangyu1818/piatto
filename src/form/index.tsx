import { useForm, useFormContext, useWatch } from 'react-hook-form'
import InternalForm from './form'
import FormItem from './form-item'

type InternalFormType = typeof InternalForm

interface FormComponent extends InternalFormType {
  Item: typeof FormItem
  useForm: typeof useForm
  useFormContext: typeof useFormContext
  useWatch: typeof useWatch
}

const Form = InternalForm as FormComponent

Form.Item = FormItem
Form.useForm = useForm
Form.useFormContext = useFormContext
Form.useWatch = useWatch

export default Form
