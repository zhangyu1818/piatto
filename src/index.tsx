export { default as Button } from './button'
export { default as Input } from './input'
export { default as Slider } from './slider'
export { default as Form } from './form'
