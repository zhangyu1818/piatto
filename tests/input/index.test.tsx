import React from 'react';
import { mount } from 'enzyme';
import Input from '../../src/input';
import { Form, FormItem } from '../../src/form';
import mountTest from '../shared/mountTest';

describe('Input', () => {
  mountTest(Input);

  it('should support input', () => {
    const wrapper = mount(<Input />);
    wrapper.find('input').simulate('change', { target: { value: '123' } });
    expect(wrapper.find('input').getDOMNode().getAttribute('value')).toBe('123');
    wrapper.find('input').simulate('change', { target: { value: '321' } });
    expect(wrapper.find('input').getDOMNode().getAttribute('value')).toBe('321');
  });

  it('should support max length', () => {
    const wrapper = mount(<Input maxLength={5} />);
    expect(wrapper.find('input').getDOMNode()).toMatchSnapshot();
  });

  it('should support focus', () => {
    const wrapper = mount(<Input />);
    wrapper.find('input').simulate('focus');
    expect(wrapper.find('.piatto-input-focus').length).toBe(1);
    wrapper.find('input').simulate('blur');
    expect(wrapper.find('.piatto-input-focus').length).toBe(0);
  });

  it('should support default value', () => {
    const wrapper = mount(<Input defaultValue="123" />);
    expect(wrapper.find('input').getDOMNode().getAttribute('value')).toBe('123');
  });

  it('should not show clear icon if default value is empty', () => {
    const wrapper = mount(<Input defaultValue="" />);
    expect(wrapper.find('.piatto-input-clear-icon').length).toBe(0);
  });

  it('should allow clear icon', () => {
    const wrapper = mount(<Input defaultValue="321" allowClear />);
    wrapper.find('.piatto-input-clear-icon').simulate('click');
    expect(wrapper.find('input').getDOMNode().getAttribute('value')).toBe('');
  });

  it('should support value be controlled', () => {
    const wrapper = mount(
      <Form>
        <FormItem name="user">
          <Input />
        </FormItem>
      </Form>,
    );
    wrapper.find('input').simulate('change', { target: { value: '123' } });
    expect(wrapper.find('input').prop('value')).toBe('123');
  });

  it('should call onChange', () => {
    const onChangeMock = jest.fn();
    const wrapper = mount(<Input onChange={onChangeMock} />);
    wrapper.find('input').simulate('change', { target: { value: '1234' } });
    expect(onChangeMock).toHaveBeenCalledTimes(1);
    expect(wrapper.find('input').getDOMNode().getAttribute('value')).toBe('1234');
  });
});
