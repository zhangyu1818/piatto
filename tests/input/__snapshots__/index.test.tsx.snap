// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should support max length 1`] = `
<input
  aria-label="Input"
  maxlength="5"
  type="text"
  value=""
/>
`;
